commit b14a2a3aea59b7f1821e118e43eba4efc2a5d6f1
Author: William Swanson <swansontec@gmail.com>
Date:   Thu Nov 5 16:05:38 2015 -0800

    ios-build.patch

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 73c9630..bb92e34 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -39,9 +39,13 @@ OPTION( USE_SSH				"Link with libssh to enable SSH support" ON )
 OPTION( USE_GSSAPI			"Link with libgssapi for SPNEGO auth"   OFF )
 OPTION( VALGRIND			"Configure build for valgrind"			OFF )
 OPTION( CURL			"User curl for HTTP if available" ON)
+OPTION( USE_OPENSSL			"Link with and use openssl library"		ON )
 
 IF(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
 	SET( USE_ICONV ON )
+ENDIF()
+
+IF(${CMAKE_SYSTEM_NAME} MATCHES "Darwin" AND NOT USE_OPENSSL)
 	FIND_PACKAGE(Security)
 	FIND_PACKAGE(CoreFoundation REQUIRED)
 ENDIF()
@@ -76,10 +80,6 @@ IF(MSVC)
 	OPTION(MSVC_CRTDBG "Enable CRTDBG memory leak reporting" OFF)
 ENDIF()
 
-IF (NOT ${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
-	OPTION( USE_OPENSSL                     "Link with and use openssl library"             ON )
-ENDIF()
-
 # This variable will contain the libraries we need to put into
 # libgit2.pc's Requires.private. That is, what we're linking to or
 # what someone who's statically linking us needs to link to.
@@ -501,8 +501,10 @@ ENDIF()
 
 # Compile and link libgit2
 ADD_LIBRARY(git2 ${SRC_H} ${SRC_GIT2} ${SRC_OS} ${SRC_ZLIB} ${SRC_HTTP} ${SRC_REGEX} ${SRC_SHA1} ${WIN_RC})
-TARGET_LINK_LIBRARIES(git2 ${SECURITY_DIRS})
-TARGET_LINK_LIBRARIES(git2 ${COREFOUNDATION_DIRS})
+IF (SECURITY_FOUND AND COREFOUNDATION_FOUND)
+	TARGET_LINK_LIBRARIES(git2 ${SECURITY_DIRS})
+	TARGET_LINK_LIBRARIES(git2 ${COREFOUNDATION_DIRS})
+ENDIF()
 TARGET_LINK_LIBRARIES(git2 ${SSL_LIBRARIES})
 TARGET_LINK_LIBRARIES(git2 ${SSH_LIBRARIES})
 TARGET_LINK_LIBRARIES(git2 ${GSSAPI_LIBRARIES})
