commit 611b2cf57668a6e58d9a587184411b3ef81cb318
Author: William Swanson <swansontec@gmail.com>
Date:   Thu Nov 5 16:08:03 2015 -0800

    pinning.patch

diff --git a/src/openssl_stream.c b/src/openssl_stream.c
index 4df7c6b..94fcac1 100644
--- a/src/openssl_stream.c
+++ b/src/openssl_stream.c
@@ -31,6 +31,88 @@
 #include <openssl/x509v3.h>
 #include <openssl/bio.h>
 
+const char *AB_OLD_AUTH_CERTIFICATE =
+	"-----BEGIN CERTIFICATE-----\n"
+	"MIIDWDCCAkACCQDt2oLI294DJTANBgkqhkiG9w0BAQsFADB0MQswCQYDVQQGEwJV\n"
+	"UzETMBEGA1UECAwKQ2FsaWZvcm5pYTESMBAGA1UEBwwJU2FuIERpZWdvMRQwEgYD\n"
+	"VQQKDAtBaXJiaXR6IEluYzEmMCQGA1UEAwwdQWlyYml0eiBDZXJ0aWZpY2F0ZSBB\n"
+	"dXRob3JpdHkwHhcNMTQwODMwMDAzNTA0WhcNMTYwMTEyMDAzNTA0WjBoMQswCQYD\n"
+	"VQQGEwJVUzETMBEGA1UECAwKQ2FsaWZvcm5pYTESMBAGA1UEBwwJU2FuIERpZWdv\n"
+	"MRQwEgYDVQQKDAtBaXJiaXR6IEluYzEaMBgGA1UEAwwRKi5zeW5jLmFpcmJpdHou\n"
+	"Y28wggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDTv4EnftlppFsRuCOg\n"
+	"PMrDEtWeD0h0s1uFh+QZ3RQxrmSavafUV2fjaXMNTV2r4kYLxz0iXHdFbZgqWlad\n"
+	"OWGKpqmCzI2mHi9fYmF7Hl+VOYD/mI5GuPhcGtiVat0lXf3WO5h1nsRPiEgvNGxs\n"
+	"0OMeyE3UktVunKZ0jIWH2AZ9yhzewhHbdnqeB9usFUDzNxxZKGu9mKR6y+VzuiGN\n"
+	"LDq4BdfV/mwgP0J73Xk35RDSDEcn9H6o9fzdol4I0vPP7YbyzSbIcB7xqEyAtC8e\n"
+	"Q0eOCjbYy8wlSgixM3saACdDWNSWbsUlqST+GoxhhrpuMwM7+BzuAq75vRwmNEPb\n"
+	"x+xNAgMBAAEwDQYJKoZIhvcNAQELBQADggEBABevQJfDE5Fjlnts4aLdWJQvhsDB\n"
+	"m+jSgsYubd0JVewjexVdM18MylcmswTymMDTn40HAy39NpoMEZXq2FT+fGflJVYG\n"
+	"p7Ra6hEg+i7uTFtcs7pLXo2mH4hWHIy3VOSm866Srlf2V6fVOExMBoQWGiIClr/Z\n"
+	"h9cDkg+nS4H42PBkCKnBYNHvaH/top4jBEUibnepYvpuoo4/evMli8T1C9fBVzpt\n"
+	"K6dA2YjCJshu7qre/48QHx4FjBO9JbQnsiWmqYkMm0XwYJazyKOI6CXRLPCc5Qxl\n"
+	"4ti9afAC70HaKNQsRMnOUlsCYKcyQ8Y0gXVPgaFBBq1BrZCfQTN6Gqo0WCo=\n"
+	"-----END CERTIFICATE-----\n";
+const char *AB_AUTH_CERTIFICATE =
+	"-----BEGIN CERTIFICATE-----\n"
+	"MIIDuDCCAqCgAwIBAgIJAO3agsjb3gMyMA0GCSqGSIb3DQEBCwUAMHQxCzAJBgNV\n"
+	"BAYTAlVTMRMwEQYDVQQIEwpDYWxpZm9ybmlhMRIwEAYDVQQHEwlTYW4gRGllZ28x\n"
+	"FDASBgNVBAoTC0FpcmJpdHogSW5jMSYwJAYDVQQDEx1BaXJiaXR6IENlcnRpZmlj\n"
+	"YXRlIEF1dGhvcml0eTAeFw0xNzA1MTkwNDA5NTdaFw0yNzA1MTcwNDA5NTdaMGgx\n"
+	"CzAJBgNVBAYTAlVTMRMwEQYDVQQIDApDYWxpZm9ybmlhMRIwEAYDVQQHDAlTYW4g\n"
+	"RGllZ28xFDASBgNVBAoMC0FpcmJpdHogSW5jMRowGAYDVQQDDBEqLnN5bmMuYWly\n"
+	"Yml0ei5jbzCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBANO/gSd+2Wmk\n"
+	"WxG4I6A8ysMS1Z4PSHSzW4WH5BndFDGuZJq9p9RXZ+Npcw1NXaviRgvHPSJcd0Vt\n"
+	"mCpaVp05YYqmqYLMjaYeL19iYXseX5U5gP+Yjka4+Fwa2JVq3SVd/dY7mHWexE+I\n"
+	"SC80bGzQ4x7ITdSS1W6cpnSMhYfYBn3KHN7CEdt2ep4H26wVQPM3HFkoa72YpHrL\n"
+	"5XO6IY0sOrgF19X+bCA/QnvdeTflENIMRyf0fqj1/N2iXgjS88/thvLNJshwHvGo\n"
+	"TIC0Lx5DR44KNtjLzCVKCLEzexoAJ0NY1JZuxSWpJP4ajGGGum4zAzv4HO4Crvm9\n"
+	"HCY0Q9vH7E0CAwEAAaNZMFcwHwYDVR0jBBgwFoAUjVTfJnh4ZcEWgt+ovyvrLihp\n"
+	"aVMwCQYDVR0TBAIwADALBgNVHQ8EBAMCBPAwHAYDVR0RBBUwE4IRKi5zeW5jLmFp\n"
+	"cmJpdHouY28wDQYJKoZIhvcNAQELBQADggEBAGgLZSdyRJr2NjA1Yrj0ZfA88w5L\n"
+	"Nn855xqhPIB4uaSKaglHiGd6Bl7ENLw4oOuathpAfdA0LTvcBUSauLf59T+CMOkn\n"
+	"YAjOSEy4To3EFmIQ0KD10lrhOo6c1OoKQ1AxggX41uJjKuaUh1DVYpUpC9gkbkxZ\n"
+	"XIHfErhqANVoOG2dRso4Bz0QQs3+1vsrFWf0ED7N+o5EMIC6DT+CMPDJbWiXje1V\n"
+	"0VhR479g+Q5Sh1YnvMUdu08rAApS+/BtqRIAun2mng2MeKH3S/iyckryhR+2Tew8\nGOIKus4RqgaV8J/aJNn259YVz3JBS81oNmCaBBNuAnh0kD5K9NvkAOhLSk8=\n-----END CERTIFICATE-----\n";
+
+static int matches_pin(X509 *cert)
+{
+	char *pem_cert = NULL;
+	BIO *b64 = NULL;
+	BUF_MEM *bptr = NULL;
+	int ret = 0;
+
+	if ((b64 = BIO_new(BIO_s_mem())) == NULL)
+	{
+		goto exit;
+	}
+	if (1 != PEM_write_bio_X509(b64, cert))
+	{
+		goto exit;
+	}
+
+	BIO_get_mem_ptr(b64, &bptr);
+
+	if (NULL == (pem_cert = malloc(bptr->length + 1)))
+	{
+		goto exit;
+	}
+	if (0 >= BIO_read(b64, pem_cert, bptr->length))
+	{
+		goto exit;
+	}
+	if (strncmp(pem_cert, AB_AUTH_CERTIFICATE, strlen(AB_AUTH_CERTIFICATE)) == 0
+	||	strncmp(pem_cert, AB_OLD_AUTH_CERTIFICATE, strlen(AB_OLD_AUTH_CERTIFICATE)) == 0)
+	{
+		ret = 1;
+	}
+exit:
+	if (pem_cert)
+		free(pem_cert);
+	if (b64)
+		BIO_free(b64);
+	return ret;
+}
+
 static int bio_create(BIO *b)
 {
 	b->init = 1;
@@ -209,6 +269,11 @@ static int verify_server_cert(SSL *ssl, const char *host)
 		return -1;
 	}
 
+	if (!matches_pin(cert)) {
+		giterr_set(GITERR_SSL, "the server certificate ");
+		return -1;
+	}
+
 	/* Check the alternative names */
 	alts = X509_get_ext_d2i(cert, NID_subject_alt_name, NULL, NULL);
 	if (alts) {
