commit 90a3f31976ef0ac1436d80003a83a3db055e9e88
Author: William Swanson <swansontec@gmail.com>
Date:   Thu Nov 5 16:27:12 2015 -0800

    timeout.patch

diff --git a/src/curl_stream.c b/src/curl_stream.c
index ca06c20..d96c93a 100644
--- a/src/curl_stream.c
+++ b/src/curl_stream.c
@@ -35,6 +35,8 @@ static int curls_connect(git_stream *stream)
 	long sockextr;
 	int failed_cert = 0;
 	CURLcode res;
+
+	curl_easy_setopt(s->handle, CURLOPT_CONNECTTIMEOUT, 10);
 	res = curl_easy_perform(s->handle);
 
 	if (res != CURLE_OK && res != CURLE_PEER_FAILED_VERIFICATION)
diff --git a/src/socket_stream.c b/src/socket_stream.c
index 71f4911..7411b85 100644
--- a/src/socket_stream.c
+++ b/src/socket_stream.c
@@ -68,6 +68,40 @@ static int close_socket(GIT_SOCKET s)
 
 }
 
+static
+int timeout_connect(int sock, struct sockaddr *addr,
+					socklen_t addr_len, struct timeval *tv)
+{
+	fd_set fdset;
+	int flags = 0;
+	int so_error;
+	socklen_t len = sizeof(so_error);
+
+	if ((flags = fcntl(sock, F_GETFL, 0)) < 0)
+		return -1;
+
+	if (fcntl(sock, F_SETFL, flags | O_NONBLOCK) < 0)
+		return -1;
+
+	if (connect(sock, addr, addr_len) == 0)
+		goto exit;
+
+	FD_ZERO(&fdset);
+	FD_SET(sock, &fdset);
+
+	if (select(sock + 1, NULL, &fdset, NULL, tv) > 0)
+	{
+		getsockopt(sock, SOL_SOCKET, SO_ERROR, &so_error, &len);
+		if (so_error == 0)
+			goto exit;
+	}
+	return -1;
+exit:
+	if (fcntl(sock, F_SETFL, flags) < 0)
+		return -1;
+	return 0;
+}
+
 int socket_connect(git_stream *stream)
 {
 	struct addrinfo *info = NULL, *p;
@@ -76,6 +110,10 @@ int socket_connect(git_stream *stream)
 	GIT_SOCKET s = INVALID_SOCKET;
 	int ret;
 
+	struct timeval sto;
+	sto.tv_sec = 10;
+	sto.tv_usec = 0;
+
 #ifdef GIT_WIN32
 	/* on win32, the WSA context needs to be initialized
 	 * before any socket calls can be performed */
@@ -111,7 +149,7 @@ int socket_connect(git_stream *stream)
 			break;
 		}
 
-		if (connect(s, p->ai_addr, (socklen_t)p->ai_addrlen) == 0)
+		if (timeout_connect(s, p->ai_addr, (socklen_t)p->ai_addrlen, &sto) == 0)
 			break;
 
 		/* If we can't connect, try the next one */
