diff --git a/src/tcp_connecter.cpp b/src/tcp_connecter.cpp
index b8d62a2..dd94a93 100644
--- a/src/tcp_connecter.cpp
+++ b/src/tcp_connecter.cpp
@@ -313,15 +313,10 @@ zmq::fd_t zmq::tcp_connecter_t::connect ()
 #ifdef ZMQ_HAVE_WINDOWS
     zmq_assert (rc == 0);
     if (err != 0) {
-        if (err != WSAECONNREFUSED
-            && err != WSAETIMEDOUT
-            && err != WSAECONNABORTED
-            && err != WSAEHOSTUNREACH
-            && err != WSAENETUNREACH
-            && err != WSAENETDOWN
-            && err != WSAEACCES
-            && err != WSAEINVAL
-            && err != WSAEADDRINUSE)
+        if (err == WSAEBADF ||
+            err == WSAENOPROTOOPT ||
+            err == WSAENOTSOCK ||
+            err == WSAENOBUFS)
         {
             wsa_assert_no (err);
         }
@@ -335,13 +330,10 @@
     if (err != 0) {
         errno = err;
         errno_assert (
-            errno == ECONNREFUSED ||
-            errno == ECONNRESET ||
-            errno == ETIMEDOUT ||
-            errno == EHOSTUNREACH ||
-            errno == ENETUNREACH ||
-            errno == ENETDOWN ||
-            errno == EINVAL);
+            errno != EBADF &&
+            errno != ENOPROTOOPT &&
+            errno != ENOTSOCK &&
+            errno != ENOBUFS);
         return retired_fd;
     }
 #endif
