--- operations.cpp	2018-11-03 09:30:36.000000000 -0700
+++ operations.new.cpp	2018-11-03 09:31:01.000000000 -0700
@@ -12,10 +12,11 @@
 
 //  define 64-bit offset macros BEFORE including boost/config.hpp (see ticket #5355) 
 #if !(defined(__HP_aCC) && defined(_ILP32) && !defined(_STATVFS_ACPP_PROBLEMS_FIXED))
-#define _FILE_OFFSET_BITS 64 // at worst, these defines may have no effect,
+// #define _FILE_OFFSET_BITS 64 // at worst, these defines may have no effect,
 #endif
 #if !defined(__PGI)
-#define __USE_FILE_OFFSET64 // but that is harmless on Windows and on POSIX
+
+// #define __USE_FILE_OFFSET64 // but that is harmless on Windows and on POSIX
       // 64-bit systems or on 32-bit systems which don't have files larger 
       // than can be represented by a traditional POSIX/UNIX off_t type. 
       // OTOH, defining them should kick in 64-bit off_t's (and thus 
@@ -26,7 +27,8 @@
       // That is required at least on Solaris, and possibly on other
       // systems as well.
 #else
-#define _FILE_OFFSET_BITS 64
+// #define _FILE_OFFSET_BITS 64
+
 #endif
 
 // define BOOST_FILESYSTEM_SOURCE so that <boost/filesystem/config.hpp> knows
