diff --git a/configure.ac b/configure.ac
index c1a2121..ddf1c21 100644
--- a/configure.ac
+++ b/configure.ac
@@ -212,12 +212,7 @@ AC_SUBST([pthread_LIBS], [-lpthread])
 
 # Require rt if on linux and output ${rt_LIBS}.
 #------------------------------------------------------------------------------
-AS_CASE([${host_os}], [*linux*],
-    [AC_CHECK_LIB([rt], [clock_gettime],
-        [AC_SUBST([rt_LIBS], [-lrt])
-         AC_MSG_NOTICE([rt_LIBS : ${rt_LIBS}])], 
-        [AC_MSG_ERROR([rt library is required but was not found.])])],
-    [AC_SUBST([rt_LIBS], [])])
+AC_SUBST([rt_LIBS], [])
 
 # Require icu-i18n of at least version 51.2 and output ${icu_i18n_CPPFLAGS/LIBS/PKG}.
 #------------------------------------------------------------------------------
